Allow hyphenated names for `rails plugin new` generator

robinboening opened the pull request:
It was not possible to create a new gem with a hyphenated name via the `rails plugin new` scaffold generator.

In reference to http://guides.rubygems.org/name-your-gem/ hyphens should be used for gems that extend other gems.

Also added two missing tests for the naming of rails plugins via the `rails plugin new` in a separate commit.

robinboening committed:
Allow hyphenated names for `rails plugin new` generator.

It was not possible to create a new gem with a hyphenated name via the `rails plugin new` generator.

The naming guide of rubygems clearly says dashes should be used for gems that extend other gems. http://guides.rubygems.org/name-your-gem/

robinboening commented:
@cappadona I had the chance to take another look into the code and force pushed some implementation changes. Locally all tests are passing.

@crankharder thanks you were right: I forgot to take care of the `bin/rails` template where `ENGINE_PATH` gets defined. This is done now. You could try again if you like.

rafaelfranca commented:
@chancancode since you were talking about this today you could review this PR :smile: :trollface:

robinboening commented:
Any news on this? :v:

chancancode commented:
Sorry for holding it up, I've been quite busy lately, but I'll try to review it next week. :(

:+1: on the idea, I wanted that just very recently. Just have the check the implementation before merging :)

robinboening commented:
Sure, thank you for your time. Have a nice weekend! :)

chancancode commented on railties/lib/rails/generators/rails/plugin/plugin_generator.rb:
```diff
@@ -50,10 +50,10 @@ def gitignore
-      template "lib/tasks/%name%_tasks.rake"
-      template "lib/%name%/version.rb"
-      template "lib/%name%/engine.rb" if engine?
+      template "lib/%underscored_name%.rb"
```
Shouldn't this be `lib/%namespaced_name%.rb`?

If you generate an engine called "foo-bar", you should `require foo/bar`, not `require foo_bar`, no?

robinboening replied:
Check

chancancode commented on railties/lib/rails/generators/rails/plugin/plugin_generator.rb:
```diff
@@ -50,10 +50,10 @@ def gitignore
-      template "lib/%name%/version.rb"
-      template "lib/%name%/engine.rb" if engine?
+      template "lib/%underscored_name%.rb"
+      template "lib/tasks/%namespaced_name%_tasks.rake"
```
ditto

robinboening replied:
This is already correct, isn't it? `foo-bar` will generate: `lib/tasks/foo/bar_tasks.rake`

chancancode commented on railties/lib/rails/generators/rails/plugin/templates/lib/tasks/%namespaced_name%_tasks.rake:
```diff
@@ -1,4 +1,4 @@
 # desc "Explaining what the task does"
-# task :<%= name %> do
+# task :<%= underscored_name %> do
```
Not sure what's the "right" thing for this, perhaps it should be rake-`namespace`'d too? (However, this is commented out, so perhaps it's not a big deal.

robinboening replied:
mhh, you are right, namespaced would be really consequent. But imho I think too much of extra code is needed to just throw in comments that 100% reflect how it might/should be used. Now it shows a usage example which is not that 100% pretty but also correct, right?

What do you think?

chancancode commented on railties/lib/rails/generators/rails/plugin/templates/rails/application.rb:
```diff
@@ -13,6 +13,6 @@
 
 Bundler.require(*Rails.groups)
-require "<%= name %>"
+require "<%= underscored_name %>"
```
see above (I think this should be `namespaced_name`)

robinboening replied:
Check

chancancode commented on railties/test/generators/plugin_generator_test.rb:
```diff
@@ -44,6 +44,21 @@ def test_invalid_plugin_name_raises_an_error
+  def test_hyphenated_plugin_name_underscores_filename_in_lib_folder
+    run_generator [File.join(destination_root, "hyphenated-name")]
+    assert_no_file "hyphenated-name/lib/hyphenated-name.rb"
+    assert_file "hyphenated-name/lib/hyphenated_name.rb", /module Hyphenated\n  module Name\n    # Your code goes here...\n  end\nend/
```
`lib/hyphenated/name.rb` (see above)

robinboening replied:
Check

chancancode commented on railties/test/generators/plugin_generator_test.rb:
```diff
@@ -44,6 +44,21 @@ def test_invalid_plugin_name_raises_an_error
+
+  def test_valid_plugin_name_not_raising_an_error
+    content = capture(:stderr){ run_generator [File.join(destination_root, "my_plugin-extension")] }
+    assert_equal "", content
```
In general, we usually avoid `assert_it_doesnt_blow_up`tests. Is there something more meaningful that we can assert on, or is this already covered by the tests below (and so it can be removed?)

robinboening replied:
Alright, I will remove these lines and add some other assertions to cover mix names with underscores and dashes.

chancancode commented:
Sorry for the delay. This is looking good. Added a few questions.

> Is there a better way than force pushing later changes? (I prefer to just keep one commit - or will you squash multiple commits together when merging?)

Either is fine.

Also, is this supposed to work with deeply nested modules? Something like `foo-bar-baz` that generates `Foo::Bar::Baz::Engine`. Seems like the code would be able to handle it just fine, but we should probably have a test for it.

robinboening commented on railties/test/generators/plugin_generator_test.rb:
```diff
@@ -252,6 +284,25 @@ def test_create_mountable_application_with_mountable_option
+    assert_file "hyphenated-name/config/routes.rb",                                          /Hyphenated::Name::Engine.routes.draw do/
+    assert_file "hyphenated-name/lib/hyphenated/name/version.rb",                            /module Hyphenated\n  module Name\n    VERSION = "0.0.1"\n  end\nend/
+    assert_file "hyphenated-name/lib/hyphenated/name/engine.rb",                             /module Hyphenated\n  module Name\n    class Engine < ::Rails::Engine\n      isolate_namespace Hyphenated\n    end\n  end\nend/
+    assert_file "hyphenated-name/lib/hyphenated_name.rb",                                    /require "hyphenated\/name\/engine"/
```
This also needs to be namespaced like this `lib/hyphenated/name.rb`

robinboening commented:
@chancancode I rebased with current master and (force) pushed the changes.

Deeply nested modules (like your example) are also working. I added a test to for a name like `deep-hyphenated-name` which passes.

The travis build is currently failing because of the previous commit (not related with this).

chancancode commented:
@robinboening thanks again for this! :smile: FYI, I fixed a few more things in 33030ea

robinboening commented:
Nice! @chancancode thanks for reviewing, merging and rounding it up. :+1:

tvdeyen commented:
🎉👍
